packages/enhanced/src/lib/container/HoistContainerReferencesPlugin.ts

```diff
@@ -3,78 +3,77 @@ import type {
   }
 
   apply(compiler: Compiler): void {
     compiler.hooks.thisCompilation.tap(
       'HoistContainerReferences',
       (compilation: Compilation) => {
-        compilation.hooks.afterOptimizeChunks.tap(
-          'HoistContainerReferences',
-          (chunks: Iterable<Chunk>) => {
-            for (const chunk of chunks) {
-              if (this.chunkContainsContainerEntryModule(chunk, compilation)) {
-                this.hoistModulesInChunk(chunk, compilation);
-              }
-            }
-          },
-        );
+        compilation.hooks.afterChunks.tap('HoistContainerReferences', () => {
+          const runtimeChunks = this.getRuntimeChunks(compilation);
+          this.hoistModulesInChunks(compilation, runtimeChunks);
+        });
       },
     );
   }
 
-  private chunkContainsContainerEntryModule(
-    chunk: Chunk,
+  private hoistModulesInChunks(
     compilation: Compilation,
-  ): boolean {
-    for (const module of compilation.chunkGraph.getChunkModulesIterable(
-      chunk,
-    )) {
-      if (module instanceof ContainerEntryModule) {
-        return true;
-      }
-    }
-    return false;
-  }
+    runtimeChunks: Set<Chunk>,
+  ): void {
+    const { chunkGraph } = compilation;
+    const partialChunk = this.containerName
+      ? compilation.namedChunks.get(this.containerName)
+      : undefined;
+
+    if (!partialChunk) return;
 
-  private hoistModulesInChunk(chunk: Chunk, compilation: Compilation): void {
-    const chunkGraph = compilation.chunkGraph;
-    const runtimeChunks = this.getRuntimeChunks(chunk, compilation);
+    const modulesToHoist = new Set<Module>();
 
-    for (const module of chunkGraph.getChunkModulesIterable(chunk)) {
-      if (
-        !chunk.hasRuntime() &&
-        (!this.containerName ||
-          (chunk.name !== this.containerName &&
-            chunk.name !== this.containerName + '_partial'))
-      ) {
-        chunkGraph.disconnectChunkAndModule(chunk, module);
+    for (const chunk of runtimeChunks) {
+      if (chunkGraph.getNumberOfEntryModules(chunk) === 0) {
+        for (const module of chunkGraph.getChunkModulesIterable(partialChunk)) {
+          if (!chunkGraph.isModuleInChunk(module, chunk)) {
+            chunkGraph.connectChunkAndModule(chunk, module);
+            modulesToHoist.add(module);
+          }
+        }
+        for (const module of modulesToHoist) {
+          chunkGraph.disconnectChunkAndModule(partialChunk, module);
+        }
       }
+    }
+
+    this.cleanUpChunks(chunkGraph, modulesToHoist);
+  }
 
-      for (const runtimeChunk of runtimeChunks) {
-        if (chunkGraph.isModuleInChunk(module, runtimeChunk)) continue;
-        chunkGraph.connectChunkAndModule(runtimeChunk, module);
+  private cleanUpChunks(
+    chunkGraph: Compilation['chunkGraph'],
+    modules: Set<Module>,
+  ): void {
+    for (const module of modules) {
+      for (const chunk of chunkGraph.getModuleChunks(module)) {
```

Comment:
remove all duplicated modules in the other entrypoints that have been hoisted into the runtimes